In [2]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
wine = load_wine()
print("Done")

Done


In [3]:
print(dir(wine))

wine.keys()

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

In [4]:
wine.data.shape

(178, 13)

In [5]:
wine.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [6]:
wine_label = wine.target
print(wine_label.shape)
wine.target

(178,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2])

In [7]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

In [8]:
x = wine.data
y = wine.target
print('Done')

Done


In [13]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 5)

In [14]:
from sklearn.tree import DecisionTreeClassifier
tree_clf = DecisionTreeClassifier()
tree_clf.fit(x_train, y_train)
y_pred = tree_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.76      0.87        17
           1       0.73      1.00      0.85        11
           2       0.88      0.88      0.88         8

    accuracy                           0.86        36
   macro avg       0.87      0.88      0.86        36
weighted avg       0.89      0.86      0.86        36



In [15]:
from sklearn.ensemble import RandomForestClassifier
rf_clf = RandomForestClassifier(random_state=11)
rf_clf.fit(x_train, y_train)
y_pred = rf_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97        17
           1       0.92      1.00      0.96        11
           2       1.00      1.00      1.00         8

    accuracy                           0.97        36
   macro avg       0.97      0.98      0.98        36
weighted avg       0.97      0.97      0.97        36



In [22]:
from sklearn.svm import SVC
svc_model = SVC()
svc_model.fit(x_train, y_train)
y_pred = svc_model.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.71      0.77        17
           1       0.69      0.82      0.75        11
           2       0.44      0.50      0.47         8

    accuracy                           0.69        36
   macro avg       0.66      0.67      0.66        36
weighted avg       0.72      0.69      0.70        36



In [23]:
from sklearn.linear_model import SGDClassifier
sgd_clf = SGDClassifier()
sgd_clf.fit(x_train, y_train)
y_pred = sgd_clf.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.59      0.74        17
           1       0.71      0.45      0.56        11
           2       0.32      0.75      0.44         8

    accuracy                           0.58        36
   macro avg       0.68      0.60      0.58        36
weighted avg       0.76      0.58      0.62        36



In [18]:
from sklearn.linear_model import LogisticRegression
lg_reg = LogisticRegression(max_iter = 10000)
lg_reg.fit(x_train, y_train)
y_pred = lg_reg.predict(x_test)

from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.94      0.97        17
           1       0.85      1.00      0.92        11
           2       1.00      0.88      0.93         8

    accuracy                           0.94        36
   macro avg       0.95      0.94      0.94        36
weighted avg       0.95      0.94      0.95        36



테스트데이터 예측 결과 해석
===================

랜덤포레스트 모형과 로지스틱리그레션 모형이 좋은 성능을 보여주고 있습니다.

랜덤포레스트 기법은 요인을 랜덤으로 선정하여 모델을 만든 후에 앙상블하는 기법인 만큼 좋은 성능을 보여주고 있습니다. 로지스틱리그레션의 경우도 라벨의 숫자가 많지 않아서인지 손글씨 데이터의 결과에 비해 좋은 성능을 보여주고 있습니다.

둘 모두 좋은 모델이며 만약 더 큰 데이터와 더 많은 라벨에 대해 적용할 경우 앙상블을 진행하는 랜덤포레스트기법이 더 좋은 성능을 내리라 예상됩니다. 랜덤포레스트기법이 가장 보편적으로 편하게 쓸 수 있는 기법이라고 생각할 수 있겠습니다.

의사결정나무는 준수하지만 랜덤포레스트나 로지스틱리그레션에 비하면 확연하게 떨어지는 성능을 보여주고 있습니다. 의사결정나무는 분류 과정에 경계선상에 있는 데이터에 대해 오분류할 가능성이 높습니다. 178개의 데이터는 의사결정나무의 오분류확률을 감내할 수 있을 정도로 많은 데이터는 아니기에 성능이 조금 낮게 나온 것이 아닌가 싶습니다. 하지만 대규모 데이터에서 실행속도가 빠르고 오분류확률이 줄어드는 만큼 대용량데이터를 대상으로 사용하기에 나쁘지 않은 모델이라고 생각됩니다.

서포트벡터머신의 성능도 낮은 수준으로 나왔습니다. 이는 와인데이터가 결정 경계로 쉽게 분류되지 않는 데이터라서 그럴 것으로 예상됩니다. 수치가 오밀조밀하거나, 분포의 모양이 전형적이지 않을 수 있다고 생각됩니다. 물론 하이퍼파라미터를 통해 모델의 성능을 더 끌어올릴 수 있겠지만 현재 노드에서 정확히 언급되지 않은 관계로 넘어가도록하겠습니다.

확률적경사하강법 모델은 와인 데이터 예측에서 가장 낮은 성능을 보여주고 있습니다. 아마 일부 데이터를 활용하여 확률을 적용하여 빠르게 최적해를 찾아내는데에 중점을 둔 모델이라 발생한 문제 같습니다. 따라서 대규모 데이터에 대해서도 분석을 진행하고 성능을 확인해 봐야 좀 더 정확한 판단이 가능하리라 생각됩니다.